tests/integration/cache_test.go

```diff
@@ -1048,3 +1112,66 @@ func collectAndAssertAtomicEvents(t *testing.T, watch clientv3.WatchChan) (event
 		}
 	}
 }
+
+func applyEvents(ctx context.Context, t *testing.T, kv clientv3.KV, evs []*clientv3.Event) int64 {
+	var lastRev int64
+	for _, batches := range batchEventsByRevision(evs) {
+		lastRev = applyEventBatch(ctx, t, kv, batches)
+	}
+	return lastRev
+}
+
+func batchEventsByRevision(events []*clientv3.Event) [][]*clientv3.Event {
+	var batches [][]*clientv3.Event
+	if len(events) == 0 {
+		return batches
+	}
+	start := 0
+	for i := 1; i < len(events); i++ {
+		if events[i].Kv.ModRevision != events[start].Kv.ModRevision {
+			batches = append(batches, events[start:i])
+			start = i
+		}
+	}
+	batches = append(batches, events[start:])
+	return batches
+}
+
+func applyEventBatch(ctx context.Context, t *testing.T, kv clientv3.KV, batch []*clientv3.Event) int64 {
+	if len(batch) == 1 {
+		event := batch[0]
+		switch event.Type {
+		case clientv3.EventTypePut:
+			resp, err := kv.Put(ctx, string(event.Kv.Key), string(event.Kv.Value))
+			if err != nil {
+				t.Fatalf("Put failed: %v", err)
+			}
+			return resp.Header.Revision
+		case clientv3.EventTypeDelete:
+			resp, err := kv.Delete(ctx, string(event.Kv.Key))
+			if err != nil {
+				t.Fatalf("Delete failed: %v", err)
+			}
+			return resp.Header.Revision
+		default:
+			t.Fatalf("Unsupported event type: %v", event.Type)
+		}
+	}
```

Comment:
This is not needed. Put/Delete by itself and Put/Delete within TXN are executed the same way.